File: leetcode/Python/maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        def _conversion(s):
            bit, t = 0, 0
            for c in s:
                t = 1<<(ord(c)-ord('a'))
                if (bit&t) != 0:
                    return
                bit|=t
            bits.append([bit, len(s)])
        sol, bits, dp = 0, [], []
        for s in arr:
            _conversion(s)
        for b in bits:
            for d in dp:
                if (b[0]&d[0]) == 0:
                    dp.append([b[0]|d[0], b[1]+d[1]])
                    sol=max(sol,dp[-1][1])
            dp.append([b[0],b[1]])
            sol=max(sol,dp[-1][1])
        return sol
        

File: leetcode/C++/maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    void conversion(string &s, vector<unsigned int> &v, vector<unsigned int> &v_lens){
        unsigned int bits=0;
        for(int i=0,bit=0; i<s.length(); ++i){
            bit=1<<(s[i]-'a');
            if((bits&bit)!=0) return;
            bits|=bit;
        }
        v.push_back(bits);
        v_lens.push_back(s.length());
    }
    int maxLength(vector<string>& arr) {
        int sol=0;
        vector<unsigned int> v,v_lens, dp, dp_lens;
        for(string &s : arr) conversion(s,v, v_lens);
        for(int i=0; i<v.size(); ++i){
            for(int j=(int)dp.size()-1; j>=0; --j){
                if((v[i]&dp[j])!=0) continue;
                dp.push_back(v[i]|dp[j]);
                dp_lens.push_back(v_lens[i]+dp_lens[j]);
                sol=max(sol,(int)dp_lens.back());
            }
            dp.push_back(v[i]);
            dp_lens.push_back(v_lens[i]);
            sol=max(sol,(int)dp_lens.back());
        }
        return sol;
    }
};